### PR TITLE
AI: improve the heuristics of covering shooters when using defensive tactics

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
1. The logic of covering shooters with units of various types has been slightly improved:

<details>
<summary>master branch</summary>

https://github.com/ihhub/fheroes2/assets/32623900/0fb42bdb-d995-4328-b10f-da8962b1189b

https://github.com/ihhub/fheroes2/assets/32623900/a1296de8-cb71-4edb-ba1e-c6b9b72049b8
</details>

<details>
<summary>this PR</summary>

https://github.com/ihhub/fheroes2/assets/32623900/fb371756-3f5f-4685-90ce-5aa1485cbc28

https://github.com/ihhub/fheroes2/assets/32623900/db75e326-ef60-48ef-9788-6c2632f841c3
</details>

Not all situations are handled perfectly - for example, in this situation, the second stack of Crusaders should have skipped a turn and let the Cavalry to cover the archers, but such behavior would have led to a significant complication of logic and is not currently implemented:

https://github.com/ihhub/fheroes2/assets/32623900/ee416d09-744f-4769-8f55-6d8f5f34205d

2. Currently in the `master` branch, melee units are more willing to cover stacks of archers closer to them, even with a large difference in the firepower of these archer stacks:

https://github.com/ihhub/fheroes2/assets/32623900/a502c11a-db87-4f5a-9ee6-a962e669e6f2

In my opinion, in a situation like this, it seems more desirable to completely cover a stronger stack than partially cover both stacks. In this PR, distance-related penalties increase more slowly, so melee units will be more willing to cover more distant, but more powerful stacks of archers:

https://github.com/ihhub/fheroes2/assets/32623900/bd89b806-bceb-432f-ab6c-11f0eb7bc3f8

In the case of stacks of comparable power, the covering forces will continue to be divided:

https://github.com/ihhub/fheroes2/assets/32623900/78248ce3-1bf6-4e8a-b0ac-239d20a280e9

If only 2 melee units are enough to cover a more powerful stack, then the second stack will also be covered:

https://github.com/ihhub/fheroes2/assets/32623900/f5024a4f-4252-4e57-abe3-dd637d16841a

### CODE REVIEW

**Review by ihhub — COMMENTED** [review]
Hi @oleg-derevenetz , I left just one question here. Can you please check it?

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 1]
```diff
@@ -1353,7 +1355,11 @@ namespace AI
 
                 return outcome.canAttackImmediately;
             } );
-            const double defenseDistanceModifier = _myArmyStrength / STRENGTH_DISTANCE_FACTOR;
+
+            // The distance that melee units need to overcome to cover the archers located on the edges of the battlefield is approximately 5 tiles. If the value of one
+            // of these archer stacks is 2 or more times greater than the value of the other, then it is worth using all melee units to create a more reliable cover for
+            // more powerful stack of archers. In other words, overcoming the distance of 5 tiles should mean the loss of one third of the total ranged units value.
+            const double defenseDistanceModifier = _myRangedUnitsOnly / 15.0;
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:1362]: Hi @ihhub OK, let's use an example. Suppose that we have two stacks of archers, one is at tile 0 with the strength of 200 and another at tile 88 with the strength of 100. Unit from tile 66 should move 1 tile to cover the second stack and (approximately) 5 tiles to cover the first stack. So `defenseDistanceModifier` in this case will be `(100 + 200) / 15 = 20`. So for the first stack, the penalty will be `20 * 5 = 100` and for the second stack the penalty will be 20. The first stack value will be `200 - 100 = 100`, and for the second stack value will be `100 - 20 = 80`, so the first stack will be preferred.

[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:1362]: @ihhub Maybe I should replace/illustrate this comment by an example of this kind?

[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:1362]: Hi @oleg-derevenetz , yes, it will help with understanding of the code greatly.

**Review by ihhub — APPROVED** [review]

**Comment by Branikolog** [comment]
Hi, @oleg-derevenetz !

>Not all situations are handled perfectly - for example, in this situation, the second stack of Crusaders should have skipped a turn and let the Cavalry to cover the archers, but such behavior would have led to a significant complication of logic and is not currently implemented:

Maybe the 3-rd troop should take the nearest available hex to the one, which cannot be reached in such situations?

defense

https://github.com/ihhub/fheroes2/assets/55348946/f5c44412-28c0-42cd-870a-e9d6fbcff0da

I think a 2-hex creature shouldn't prefer bottom position while covering shooters on the right side of the battlefield.

**Comment by oleg-derevenetz** [comment]
Hi @Branikolog

> Maybe the 3-rd troop should take the nearest available hex to the one, which cannot be reached in such situations?

Even if it is, it is out of scope of this PR. I would not like to produce "special cases", but the general solution is a bit complicated: ideally, we need to select positions that do not allow all (or at least most) enemy units to approach, and such a calculation in general looks expensive, because we will need to go through a lot of variants and recalculate the pathfinder cache many times, so the logic is currently limited to cells directly adjacent to the shooting units.

**Comment by oleg-derevenetz** [comment]
> I think a 2-hex creature shouldn't prefer the bottom position while covering shooters on the right side of the battlefield.

Now the logic needs to be improved in this regard.

**Review by Branikolog — APPROVED** [review]
I@oleg-derevenetz 
Now it works fine!

I've noticed a strange bahaviour. In a defensive situation troops sometimes decide to attack the nearest enemy. But sometimes not.

https://github.com/ihhub/fheroes2/assets/55348946/17e3256e-a7de-4478-bfe7-2ee7a6024a1b

https://github.com/ihhub/fheroes2/assets/55348946/2cb05629-f095-49a9-a94e-bafd91eb8385

I suppose it depends on whether the enemy troop blocks path of the defending cavalry.

But sometimes defending cavalry walks around, while enemy troop definitely blocks the way to the shooters.

https://github.com/ihhub/fheroes2/assets/55348946/0e456002-3063-46a7-86eb-1da49f86d18d

Do they act in accordance with the existing logic or something went wrong?

**Comment by oleg-derevenetz** [comment]
@Branikolog

> In a defensive situation troops sometimes decide to attack the nearest enemy. But sometimes not.

If unit cannot approach any of friendly shooter stacks in 2 turns, but it can attack some enemy unit on the "friendly" half of the battlefield on the current turn, then it will attack that unit, because since there are already enemy units on the "friendly" half of the battlefield, then most likely they will be able to approach friendly shooters first anyway, so it's better to try to make damage to them or even kill.

**Comment by Branikolog** [comment]
@oleg-derevenetz 

Ah, 2 turns rule. Now it's clear. Thank you for the explanation! :)

**Comment by oleg-derevenetz** [comment]
@Branikolog

> Ah, 2 turns rule. Now it's clear. Thank you for the explanation! :)

BTW since the covering logic is now improved, we can try to turn it to the "1 turn rule".

**Comment by ihhub** [comment]
@oleg-derevenetz , thanks a lot for this improvement!

**Comment by Branikolog** [comment]
Hi, @ihhub 

> > Ah, 2 turns rule. Now it's clear. Thank you for the explanation! :)
> 
> BTW since the covering logic is now improved, we can try to turn it to the "1 turn rule".

Only if defending troops would try to position themselves as close to the shooter being protected as possible.

**Comment by oleg-derevenetz** [comment]
> Only if defending troops would try to position themselves as close to the shooter being protected as possible.

Well, yes, it needs to be thought out separately.

**Pull request merged**